extensions/data-plane/data-plane-http/src/main/java/org/eclipse/dataspaceconnector/dataplane/http/pipeline/HttpDataSink.java

```diff
@@ -0,0 +1,147 @@
+/*
+ *  Copyright (c) 2021 Microsoft Corporation
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+package org.eclipse.dataspaceconnector.dataplane.http.pipeline;
+
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import org.eclipse.dataspaceconnector.dataplane.spi.pipeline.DataSink;
+import org.eclipse.dataspaceconnector.dataplane.spi.pipeline.DataSource;
+import org.eclipse.dataspaceconnector.dataplane.spi.result.TransferResult;
+import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+
+import static java.lang.String.format;
+import static java.util.concurrent.CompletableFuture.allOf;
+import static java.util.concurrent.CompletableFuture.supplyAsync;
+import static java.util.stream.Collectors.toList;
+import static org.eclipse.dataspaceconnector.spi.transfer.response.ResponseStatus.ERROR_RETRY;
+
+/**
+ * Writes data in a streaming fashion to an HTTP endpoint.
+ */
+public class HttpDataSink implements DataSink {
+    private String authKey;
+    private String authCode;
+    private String endpoint;
+    private String requestId;
+    private OkHttpClient httpClient;
+    private ExecutorService executorService;
+    private Monitor monitor;
+
+    @Override
+    public CompletableFuture<TransferResult> transfer(DataSource source) {
+        try {
+            var futures = source.openPartStream().map(part -> supplyAsync(() -> postData(part), executorService)).collect(toList());
+            return allOf(futures.toArray(CompletableFuture[]::new)).thenApply((s) -> {
+                if (futures.stream().anyMatch(future -> future.getNow(null).failed())) {
+                    return TransferResult.failure(ERROR_RETRY, "Error transferring data");
+                }
+                return TransferResult.success();
+            });
```

Comment:
Maybe it's an idea to add here some exception handling. I assume if one requests fails, all remaining requests are still send to the other participant.

If a lot of part requests are send to the other participant, are they all executed at the same time? Maybe this could cause a timeout on some of the requests.